### ISSUE TITLE
Middleware that inspects request and passes it on to innerhandler causes exception.

### ISSUE THREAD

**Issue by jonaskello** [dart-lang/shelf#9]
I'm trying to create a middleware in Shelf that will inspect the request, and if certain values are found the request will be handled, otherwise it should be sent to the inner handler. Eg. I would like to inspect the Request.method.

 ```dart
Handler middleware(Handler innerHandler) {

    return (Request req) {
        if(req.method == "GET" && req.headers["xxx"] == yyy) {
            // Handle the request
            ...
        }
        else {
            // This gives exception:
            // Bad state: The 'read' method can only be called once on a shelf.Request/shelf.Response object.
            return innerHandler(req);
        }
    }
 ```

The problem is that it is not possible to call the inner handler after the Request has been inspected. How do I go about inspecting it but still being able to send it along to the inner handler?

**Comment by nex3** [comment]
I can't reproduce this with [this test app](https://gist.github.com/nex3/7186ccee586b8e5fef21):

```shell
$ dart app.dart &
[1] 12007
$ curl http://localhost:1234
success
```

What's different between what you're doing and that test app?

**Comment by jonaskello** [comment]
I tried your gist and it works. My real code is much more complex and has middleware for authentication, proxy etc. I extracted the part I thought was causing trouble and put in my example but it seems like you have proven it is not the actual problem. I then went through all the code and cut it down piece by piece to the minimal to reproduce the error. I put the gist for that minimal reproduction here:

https://gist.github.com/jonaskello/ebba480af1808530ef15

I think it might be related to shelf_proxy rather than shelf itself?

**Comment by kevmoo** [comment]
@jonaskello I'm looking into it...

**Comment by kevmoo** [comment]
@jonaskello The proxy is called twice in the cascade. Each time, it's read by the proxy to be able to make the request. The second call causes the error.

This is by design, since requests can be streamed and we don't want to hold on to a copy of the contents.

**Comment by jonaskello** [comment]
I should probably have been more specific about my use case. I have implemented a client-side router that supports pushstate. In order for this to work the shelf server needs to rewrite all requests to index.html unless the request is for a real existing file (because the client side-router may have an URL that does not exists but that will work as long as the server always returns the SPA, which is index.html).

The reason I use shelf_proxy is because I am using Webstorm and I need to go through it's built-in server in order for debugging to work. So what I am trying to do is setup the shelf server to do this:

Request comes in -> shelf_proxy to webstorm's server, if status != 404 then serve it -> if status == 404 then continue to my "rewriteToIndex" middleware which will change the requested path to "/index.html" and then call the inner handler  -> The inner handler is again shelf_proxy which in will proxy to Webstorm wihich will now find "/index.html" and serve it.

The first step is needed because in addition to index.html there are other static files that needs to be served. My middleware is very similar to this one for node:

https://github.com/bripkens/connect-history-api-fallback

So the qustion becomes how to support this scenario in shelf if the request cannot be read twice?

**Comment by kevmoo** [comment]
@jonaskello I think you'll need to copy the content of the Request into a byte-array and re-create the request as needed.

You could imagine a `CopiedRequest` class that contains the `Request` content in a static form, which could support reading multiple times.

**Comment by jonaskello** [comment]
@kevmoo I created a class CopiedRequest that extends Request, keeps an internal buffer for the body as List<List<int>>, and overrides read() to return a new Stream over the buffer. I also added a method that fully reads the body. Then, as the first middleware, I created a middleware that creates a CopiedRequest from the Request, and awaits the method that fully reads the body. Then it passes along CopiedRequest to inner handler which means all inner handlers will use the CopiedRequest instead of the original Request and thus can read the body at any point they want. It seems to work but not sure that was what you meant? 

Also it seems like a lot of work and a brittle solution as I have no idea what all of the Request and Message stuff are doing (hijack etc.). I think it would be nice if Request itself had a method createBufferedRequest() that returned a BufferedRequest that had the body buffered and could be read multiple times. There could also be a standard middleware to turn the Request into a BufferedRequest, something like bufferRequestMiddleware.
